!htt #18919 #19519 Align Java HTTP server DSL with Scala

ktoso opened the pull request:
Taking over the new Java DSL over the finish line, based on the excellent work by Jan @jypma from PR #19632.

Resolves #19519 
Resolves #19632 
Allows for basis to reviving better (and complete) JSON support as well.

> This replaces the Java HTTP server DSL with a Java-8 centric one which exposes all scala DSL concepts to be usable from Java, including custom directives, (un)marshallers, rejections, headers, and type safety for path and query parameters. 
> 
> Implemented according to the direction agreed upon with @ktoso in https://github.com/akka/akka/issues/18919 . The class `JavaRouteTest` demoes the Java DSL's capabilities.
> 
> List of things to be done (somewhat in order):
> - [X] Migrated extra directives: `FileAndResourceDirectives`, `CodingDirectives`,`CookieDirectives`,`MiscDirectives`,  `HostDirectives`, `SchemeDirectives`, `WebSocketDirectives`, `CacheConditionDirectives`,  `RangeDirectives`
> - [X] Migrate the existing tests in `akka-http-tests` once the DSL direction is agreed 
> - [X] Fix compile errors in `akka-docs`
> - [X] Fill in the missing directives that haven't been migrated yet 
> - [x] Discuss how to resolve the introduced `TODO` and `FIXME` markers
> - [ ] Replace the current Java `RouteTestKit` (which has its own implementation different from the scala `RouteTestKit`) with a [wrapper](https://github.com/jypma/scala-japi/blob/master/akkaStreamsTest/src/main/scala/com/tradeshift/scalajapi/akka/route/RouteTestKit.scala) that delegates and re-uses the scala one.
> - [ ] Update textual documentation in `akka-docs`
> - [x] MUST: Add `FormFieldDirectives`
> - [x] MUST: Add `SecurityDirectives`, possibly Java-model a bit of the related classesupon
> - [x] MUST: Add `DebuggingDirectives`
> - [x] MUST: Add `FileUploadDirectives`
> - [x] MUST: Add `RespondWithDirectives`

jypma committed:
!htt #18919 #19519 Align Java HTTP server DSL with Scala

This commits replaces the Java HTTP server DSL with a Java-8 centric one
which exposes all scala DSL concepts to be usable from Java, including
custom directives, (un)marshallers, rejections, headers, and type safety
for path and query parameters.

jypma commented:
`akka-http-jackson-experimental: Binary compatibility check failed!` : that was to be expected, marshalling has changed completely.

But:

`akka-http-testkit: Binary compatibility check failed!` Why is testkit not experimental? The routes underneath that it's testing are experimental as well. And a testkit is needed for the new DSL.

ktoso commented:
We went a bit too extreme on enabling MiMa recently, don't worry about it.
Also, it will make writing the migration guide a bit easier, so we don't miss anything.

ktoso commented:
I'll do a consistency spec as next step

ktoso committed:
WIP on consistency spec

ktoso commented:
FYI, around 44 missing directives still. (though my consistency spec may be off).

ktoso committed:
WIP on adding missing directives

jypma commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/BasicDirectives.scala:
```diff
@@ -128,6 +131,13 @@ abstract class BasicDirectives {
   /**
+   * Extracts a single value using the given function.
+   */
+  def extract[T](extract: JFunction[RequestContext, T], inner: JFunction[T, Route]): Route = ScalaRoute {
```
are you actually planning to extract a whole Java requestcontext? I didn't do that, since it would've added a lot of Java work (that part is really scala-only). I couldn't see any real use cases for it, even though it would of course be needed if you want 100% feature parity. It would also require introducing RouteResult on the Java side.

ktoso replied:
> are you actually planning to extract a whole Java requestcontext? I didn't do that, since it would've added a lot of Java work (that part is really scala-only).

As library authors we take all the pain ourselves such that users don't have to :wink:
It takes a special kind of masochism (and / or dedication) to care about the Java API being complete :wink: 

> I couldn't see any real use cases for it, even though it would of course be needed if you want 100% feature parity. It would also require introducing RouteResult on the Java side.

It is a pretty core concept and would be weird to skip now that I'm browsing things.

https://github.com/akka/akka/blob/master/akka-http/src/main/scala/akka/http/javadsl/server/RequestContext.scala

https://github.com/akka/akka/pull/20088/files#diff-fe1dc8b93e367e290481ad720726ef32L17

ktoso commented on akka-actor-tests/src/test/java/akka/util/ByteStringTest.java:
```diff
@@ -20,7 +20,7 @@ public void testCreation() {
     final ByteStringBuilder sb = ByteString.createBuilder();
     sb.append(ByteString.fromString("Hello"));
-    sb.append(ByteString.fromString("World"));
+    sb.append(ByteString.fromString("xxx"));
```
fun fact, we don't run javatests so this did not fail :O

ktoso committed:
WIP

ktoso committed:
WIP

jypma commented:
@ktoso I'm back; if I have some time, shall I put a commit somewhere to do an attempt at a Java RequestContext? Do you want all of RouteSettings and the other settings java-lized as well?

ktoso commented:
That'd be awesome if you could attack the RequestContext :) I'll focus on this PR today as well. 
You can fork off from my PR and PR towards it (inception!)

jypma committed:
Add RequestContext and RouteResult to Java DSL

ktoso committed:
massive progress

ktoso committed:
The javadsl seems complete ish

ktoso committed:
fix websockets

patriknw commented:
We decided to not rush this into today's release of 2.4.3. We will complete it as soon as possible and then release another 2.4.x. Thanks for all hard work!

jypma committed:
WIP bring java docs up to date.

This applies some updates to the root-level documentation

ktoso committed:
Merge pull request #7 from jypma/docs

WIP bring java docs up to date.

jypma committed:
[htp] Fix java documentation to correctly mention timeouts

Timeouts are configured the same in Java and Scala. Hence, linking to the
scala docs for timeouts from Java.

ktoso committed:
Merge pull request #8 from jypma/docs

[htp] Fix java documentation to correctly mention timeouts

ktoso committed:
=htc fix optionalHeaderValueByType in Java

ktoso committed:
=htt #20200 fix java testkit always using NoLogging instead logger

ktoso committed:
+htt actually run new javadsl tests, allow overriding config

ktoso committed:
=htt improve javadsl test infra with more details when fails

ktoso committed:
=htt improve javadsl test infra with more details when fails

ktoso committed:
=htt fix bug in wrong path matcher exposed

ktoso committed:
+htp add missing remaining path matcher

ktoso committed:
WORK IN PROGRESS, NOT PASSING TESTS yet COMPILING

ktoso committed:
Merge branch 'jypma-java-http-server-dsl' of github.com:ktoso/akka into jypma-java-http-server-dsl

ktoso committed:
revert

johanandren committed:
Cookie tests fixed

ktoso committed:
Merge pull request #10 from johanandren/wip-java-cookies-johanandren

Cookie tests fixed

johanandren committed:
ParameterDirectivesTest fixed

johanandren committed:
Protect the tweets from scalariform

ktoso committed:
Merge pull request #11 from johanandren/wip-java-parameter-test-johanandren

ParameterDirectivesTest fixed

ktoso committed:
WIP on path directives

johanandren committed:
Incorrect response expectations in cache condition directives spec fixed

ktoso committed:
+!htt PathMatchers rewritten, made uniform and tests passing

ktoso committed:
Merge pull request #12 from johanandren/wip-cache-condition-directives-spec-johanandren

Incorrect response expectations in cache condition directives spec fixed

johanandren committed:
Bugfix in java reject and a little test-boyscouting

ktoso committed:
Merge pull request #13 from johanandren/wip-execution-directives-test-johanandren

Bugfix in java reject and a little test-boyscouting

ktoso committed:
Revert "Incorrect response expectations in cache condition directives spec fixed"

This reverts commit cd50e89d45db010309f8249b090ea654ebb11c7a.

johanandren committed:
Fix for cache condition spec not working

* Not sending in Unit instad of the implicit magnet in the test
* HeaderMagnet now works as expected
* Java API added for - and + on DateTime

johanandren committed:
HttpAPIsTest is compile time only, not for running

Also, moved from the client package since not strictly a client test.

johanandren committed:
RequestValTest sprinkled out across multiple directive tests

Client ip extraction test with incorrect header name fixed.

ktoso committed:
Merge pull request #16 from johanandren/wip-replace-and-fix-request-val-test-johanandren

RequestValTest sprinkled out across multiple directive tests

johanandren committed:
SecurityDirectives passing

Two faulty tests and two actual bugs.

ktoso committed:
Merge pull request #17 from johanandren/wip-security-directives-johanandren

SecurityDirectives passing

johanandren committed:
PetStore example and test fixed

* Annotations to make marshalling work without default constructor
* Made model class immutable

ktoso committed:
Merge pull request #14 from johanandren/wip-cache-condition-fix-for-realz-johanandren

Fix for cache condition spec not working

ktoso committed:
Merge pull request #15 from johanandren/wip-fix-http-api-tests-johanandren

HttpAPIsTest is compile time only, not for running

ktoso committed:
Merge pull request #18 from johanandren/wip-pet-store-api-test-johanandren

PetStore example and test fixed

johanandren committed:
Incorrect tests fixed

Some scaladoc boyscouting as bonus

ktoso committed:
Merge pull request #19 from johanandren/wip-handler-example-doc-test-johanandren

Incorrect tests fixed

johanandren committed:
Incorrect CodingDirectivesTest fixed.

ktoso committed:
Merge pull request #20 from johanandren/wip-coding-directives-test-johanandren

Incorrect CodingDirectivesTest fixed.

johanandren committed:
Bugfix for Java Unmarshaller.firstOf and fixes to JavaRouteTest

ktoso committed:
Merge pull request #21 from johanandren/wip-java-route-test-johanandren

Bugfix for Java Unmarshaller.firstOf and fixes to JavaRouteTest

johanandren committed:
Thread.sleep in default exception handler removed

ktoso committed:
Merge pull request #22 from johanandren/wip-route-directive-test-johanandren

Thread.sleep in default exception handler removed

johanandren committed:
MarshallerTest fixed

* Missing seal signature added to JavaDSL
* More consistent (with Scala) test kit setup for Java
* missing Javadocs added

ktoso committed:
Merge pull request #23 from johanandren/wip-marshaller-test-johanandren

MarshallerTest fixed

ktoso committed:
copy directive docs, prepare for finishing it up

johanandren committed:
SecurityDirectives.authorize variants and test coverage added

ktoso committed:
Merge pull request #24 from johanandren/wip-complete-java-security-directives-johanandren

SecurityDirectives.authorize variants and test coverage added

johanandren committed:
Custom headers in Java DSL

johanandren committed:
Make the header "weird"

johanandren committed:
How about them apples

johanandren committed:
Restructured them apples a bit cleaner

ktoso committed:
WIP on java docs

ktoso committed:
Merge pull request #25 from johanandren/wip-custom-headers-johanandren

WIP Custom headers in Java DSL

ktoso committed:
WIP

ktoso committed:
+htp add missing parameterOrDefault directive

johanandren committed:
Work so far

ktoso committed:
WIP

johanandren committed:
More work

ktoso committed:
Merge pull request #26 from johanandren/wip-java-directives-docs-johanandren

WIP Work on half of the java directives docs

johanandren committed:
Moar work

johanandren committed:
And some moar

ktoso committed:
Merge pull request #28 from johanandren/wip-java-directives-docs-2-johanandren

Wip java directives docs 2 johanandren

ktoso committed:
WIP

ktoso committed:
Merge pull request #27 from ktoso/wip-java-directive-docs

Wip java directive docs

ktoso committed:
Link to issue about docs and fix compile error

johanandren committed:
Fixed a lot of doc warnings

ktoso committed:
Merge pull request #30 from johanandren/wip-fixing-warnings-johanandren

Fixed a lot of doc warnings

ktoso committed:
compile, migration guide

ktoso committed:
Merge pull request #29 from ktoso/wip-java-directive-docs

Link to issue about docs and fix compile error
